### PR TITLE
test: Improving credit transfer, transfer account tests

### PR DESCRIPTION
**Pull request by tristanhcole**
**Describe the Pull Request**
Adding some unit tests to the credit transfer utils file. Adding functional transfer account tests for the API (which is currently untested).

Also fixed and simplified `transfer_account.approve_and_disburse()`. Wrote tests and did manual QA.

**Todo**

- ~~[ ] Link relevant [trello card](https://trello.com/b/PA2LhlOh/sempo-dev) or [related issue](https://github.com/teamsempo/SempoBlockchain/issues) to the pull request~~
- [ ] Request review from relevant @person or team
- [ ] QA your pull request. This includes running the app, login, testing changes etc.
- [ ] Bump [backend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/config.py#L7) and/or [frontend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/app/package.json#L3) versions following Semver

### CODE REVIEW

File: app/test_app/conftest.py [hunk 1]
```diff
@@ -106,6 +106,17 @@ def create_transfer_account_user(test_client, init_database, create_organisation
...
 
+@pytest.fixture(scope='function')
+def create_transfer_account_user_function(test_client, init_database, create_organisation):
+    from server.utils.user import create_transfer_account_user
+    user = create_transfer_account_user(first_name='Transfer',
+                                        last_name='User 2',
+                                        phone=fake.msisdn(),
+                                        organisation=create_organisation)
+    db.session.commit()
+    return user
```
Comment (tristanhcole) [inline app/test_app/conftest.py:117]: Using `create_transfer_account_user` was causing problems in other tests due to the fixture being scoped to a `module`. This fixture is thus scoped to a `function`.

File: app/test_app/functional/api/test_transfer_account_api.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+import pytest
+from server.utils.auth import get_complete_auth_token
+
+
+@pytest.mark.parametrize("transfer_account_id_accessor, tier, status_code", [
+    (lambda u: u.transfer_account.id, None, 401),
+    (lambda u: u.transfer_account.id, 'superadmin', 201),  # todo(#320): this should be a 200 Status code
+    (lambda u: u.transfer_account.id, 'view', 201),  # todo(#320): this should be a 200 Status code
+    (lambda o: 1222103, 'superadmin', 400),  # todo(#320): this should be a 404 Status code
```
Comment (tristanhcole) [inline app/test_app/functional/api/test_transfer_account_api.py:9]: adding todo's for #320

File: app/test_app/functional/api/test_user_api.py [hunk 3]
```diff
@@ -79,12 +82,17 @@ def test_create_user(test_client, authed_sempo_admin_user, init_database, create
...
         assert data['user']['location'] == 'Elwood'
         assert data['user']['business_usage_id'] == init_database.session.query(TransferUsage)\
             .filter_by(name=business_usage_name).first().id
         assert data['user']['referred_by'] == user_phone_accessor(create_transfer_account_user)
 
+        if initial_disbursement is not None:
+            assert data['user']['transfer_accounts'][0]['balance'] == initial_disbursement
+        else:
+            db_user = init_database.session.query(User).get(data['user']['id'])
+            assert data['user']['transfer_accounts'][0]['balance'] == db_user.default_organisation.default_disbursement
```
Comment (tristanhcole) [inline app/test_app/functional/api/test_user_api.py:94]: adds a test for both usecase's when a user with default disbursement is created and when a custom disbursement is created (in this case, the amount of zero)

File: app/server/api/transfer_account_api.py [hunk 4]
```diff
@@ -168,7 +166,7 @@ def put(self, transfer_account_id):
 transfer_account_blueprint.add_url_rule(
     '/transfer_account/',
     view_func=TransferAccountAPI.as_view('transfer_account_view'),
-    methods=['GET', 'PUT', 'POST'],
```
Comment (tristanhcole) [inline app/server/api/transfer_account_api.py:171]: There is no `POST` method for this API

File: app/server/api/transfer_account_api.py [hunk 5]
```diff
@@ -49,9 +47,9 @@ def get(self, transfer_account_id):
             base_query = TransferAccount.query.filter(TransferAccount.is_ghost != True)
 
             if account_type_filter == 'vendor':
-                transfer_accounts_query = base_query.filter_by(has_vendor_role=True)
+                transfer_accounts_query = base_query.filter_by(is_vendor=True)
             elif account_type_filter == 'beneficiary':
-                transfer_accounts_query = base_query.filter_by(has_vendor_role=False)
+                transfer_accounts_query = base_query.filter_by(is_beneficiary=True)
```
Comment (tristanhcole) [inline app/server/api/transfer_account_api.py:52]: this was causing an error when I wrote a test for it so change to `is_vendor` and `is_beneficiary`

File: app/test_app/unit/models/test_transfer_account.py [hunk 6]
```diff
@@ -18,6 +19,7 @@ def test_approve_beneficiary_transfer_account(create_transfer_account_user, crea
     THEN check a BENEFICIARY is disbursed initial balance
     """
     create_transfer_account_user.transfer_account.is_beneficiary = True
+    g.active_organisation = create_organisation
```
Comment (tristanhcole) [inline app/test_app/unit/models/test_transfer_account.py:22]: unit test needs `g.active_organisation` to be manually set

File: app/server/models/transfer_account.py [hunk 7]
```diff
@@ -188,39 +188,30 @@ def get_approval(self, receiving_address):
...
+            # initial disbursement defaults to None. If initial_disbursement is set then skip this section.
+            # If none, then we want to see if the active_org has a default disbursement amount
+            initial_disbursement = active_org.default_disbursement
         if not initial_disbursement:
+            # if initial_disbursement is still none, then we don't want to create a transfer.
             return None
 
+        if initial_disbursement == active_org.default_disbursement:
+            auto_resolve = True
+
```
Comment (tristanhcole) [inline app/server/models/transfer_account.py:214]: There is a potential issue here where a user who is part of `<ORG1>` and `<ORG2>` is approved from the `active_org` (`default_disbursement=100`) when the user's `default_organisation` actually has a different `default_disbursement`. Would love your thoughts @enjeyw @michielderoos

Comment (tristanhcole) [reply app/server/models/transfer_account.py:214]: Maybe not an ISSUE, but a potential side effect that needs to be clear to end-users.

File: app/test_app/conftest.py (hunk 1 above)
Comment (enjeyw) [reply app/test_app/conftest.py:118]: is it viable to just scope the original `create_transfer_account_user` to a function? Broadly speaking i think it should be in most cases anyways...

File: app/test_app/functional/api/test_transfer_account_api.py (hunk 2 above)
Comment (enjeyw) [reply app/test_app/functional/api/test_transfer_account_api.py:9]: any reason not to just change them in this pr?

File: app/server/models/transfer_account.py [hunk 8]
```diff
@@ -188,39 +188,30 @@ def get_approval(self, receiving_address):
...
-        auto_resolve = initial_disbursement == active_org.default_disbursement
 
         if not self.is_approved and admin and AccessControl.has_sufficient_tier(admin.roles, 'ADMIN', 'admin'):
             self.is_approved = True
 
         if self.is_beneficiary:
-            # TODO: make this more robust
-            # approve_and_disburse might be called for a second time to disburse
-            # so first check that no credit transfer have already been received
-            if len(self.credit_receives) < 1:
```
Comment (enjeyw) [inline app/server/models/transfer_account.py:203]: i can't see where this logic is re-implemented?

File: app/test_app/functional/api/test_transfer_account_api.py (hunk 2 above)
Comment (tristanhcole) [reply app/test_app/functional/api/test_transfer_account_api.py:9]: Trying to keep each PR [atomic](https://en.wikipedia.org/wiki/Atomic_commit)

File: app/test_app/conftest.py (hunk 1 above)
Comment (tristanhcole) [reply app/test_app/conftest.py:118]: I think doing so caused some errors to occur in other tests - I will investigate

File: app/server/models/transfer_account.py (hunk 8 above)
Comment (tristanhcole) [reply app/server/models/transfer_account.py:217]: I checked the code and `approve_and_disburse` can only really be called once unless someone _manually_ set `is_approved` to False.

Comment (tristanhcole) [reply app/server/models/transfer_account.py:217]: Additionally, the approval function is broken on master branch. This fixes and this simplifies the process.

File: app/test_app/functional/api/test_metrics_api.py [hunk 9]
```diff
@@ -233,11 +233,11 @@ def get_metrics(metric_type):
         total_spent_val = 25
 
     elif metric_type == 'credit_transfer' or metric_type == 'all':
-        assert returned_stats['daily_disbursement_volume']['aggregate']['ORGANISATION'] == 300
+        assert returned_stats['daily_disbursement_volume']['aggregate']['ORGANISATION'] == 700
         assert returned_stats['daily_transaction_volume']['aggregate']['USER'] == 150
         assert returned_stats['exhausted_balance'] == 0
         assert returned_stats['has_transferred_count'] == 2
-        assert returned_stats['total_distributed'] == 300
+        assert returned_stats['total_distributed'] == 700
```
Comment (tristanhcole) [inline app/test_app/functional/api/test_metrics_api.py:240]: **Disbursement total is now 700**, as original 300 was the custom disbursement in lines 29 and 37 of this file and 400 is the default disbursement amount of `create_organisation`

Comment (tristanhcole) [reply app/test_app/functional/api/test_metrics_api.py]: This is outdated but still applies in the metrics JSON objects

File: app/server/models/transfer_account.py (hunk 8 above)
Comment (enjeyw) [reply app/server/models/transfer_account.py:217]: hmmmm i might review this overall.

**Pull request merged**